kotlinx-coroutines-core/common/src/CancellableContinuation.kt

```diff
@@ -201,6 +201,11 @@ public interface CancellableContinuation<in T> : Continuation<T> {
     public fun resume(value: T, onCancellation: ((cause: Throwable) -> Unit)?)
 }
 
+internal fun <T> CancellableContinuation<T>.invokeOnCancellation(handler: CancelHandler) = when (this) {
+    is CancellableContinuationImpl -> invokeOnCancellationInternal(handler)
+    else -> invokeOnCancellation(handler::invoke)
```

Comment:
Could you please elaborate on what code path hits this branch?

Reply:
The code path where someone implemented their own `CancellableContinuation`, against all odds. I *could* throw an exception here instead just to be mean, but why?

Reply:
In a few places, we already rely on the fact that `CancellableContinuation` can be casted to `CancellableContinuationImpl` and, I can bet that, our other codepaths are neither ready nor tested for customly implemented `CC` instances. I would prefer fail fast here.

Also, added `CancellableContinuation` to https://github.com/Kotlin/kotlinx.coroutines/issues/3770

Reply:
If we don't support third-party implementations, this would mean we can also remove `invokeOnCompletion(onCancelling, invokeImmediately, handler)`. Except ktor apparently uses it: https://grep.app/search?q=invokeOnCompletion%28onCancelling

Reply:
Not sure I follow, could you please elaborate?
One is the implementation of the interface that e.g. doesn't work with channels and mutex, and the other is the use of an internal API (with no public alternatives).

Reply:
It was about `Job`, and it turns out that branch is still needed, so the point is moot.